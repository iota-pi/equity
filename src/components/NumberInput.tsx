import React, { Component, ChangeEvent, KeyboardEvent } from "react";
import TextField from "@material-ui/core/TextField";
import InputAdornment from "@material-ui/core/InputAdornment";
import People from "@material-ui/icons/People";

export interface Props {
  value?: number,
  onChange?: (arg: number) => void,
  onEnterPress?: () => void,
  label?: string,
  variant?: "filled" | "outlined" | "standard",
  min?: number,
  max?: number,
  step?: number,
  icon?: Component,
  fullWidth?: boolean,
  autoFocus?: boolean,
}

class TextInput extends Component<Props> {
  state = {
    value: 1
  };

  render() {
    return (
      <TextField
        autoFocus={this.props.autoFocus}
        label={this.props.label}
        type="number"
        {...this.variantProps()}
        fullWidth={this.props.fullWidth}
        onChange={this.handleChange}
        onKeyPress={this.handleKeyPress}
        value={this.state.value.toString()}
        InputProps={{
          inputProps: {
            min: this.props.min || 1,
            max: this.props.max,
            step: this.props.step || 1,
          },
          startAdornment: (
            <InputAdornment position="start">
              {this.props.icon || <People />}
            </InputAdornment>
          )
        }}
      />
    );
  }

  private handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const newValue = parseInt(event.target.value);
    this.setState({ value: newValue });
    if (this.props.onChange) {
      this.props.onChange(newValue);
    }
  };

  private handleKeyPress = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') {
      if (this.props.onEnterPress) {
        this.props.onEnterPress();
      }
    }
  }

  componentDidUpdate (prevProps: Props) {
    // Update value in state if prop has changed (and prop value doesn't already match state value)
    const { value } = this.props;
    if (value !== undefined && prevProps.value !== value && value !== this.state.value) {
      this.setState({ value })
    }
  }

  private variantProps () {
    if (this.props.variant === 'filled') {
      return {variant: 'filled' as 'filled'}
    }
    if (this.props.variant === 'outlined') {
      return {variant: 'outlined' as 'outlined'}
    }
    return {variant: 'standard' as 'standard'}
  }
}

export default TextInput;
