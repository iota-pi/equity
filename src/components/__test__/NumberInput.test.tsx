import React from 'react';
import NumberInput from '../NumberInput';
import { shallow, mount, render } from 'enzyme';

describe('<NumberInput/>', () => {
  it('renders consistently', () => {
    const wrapper = render(
      <NumberInput value={1} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('responds to input correctly', () => {
    const change = jest.fn();
    const wrapper = mount(
      <NumberInput value={1} onChange={change} />
    );

    expect(change).toHaveBeenCalledTimes(0);
    wrapper.find('input').first().simulate('change', {
      target: { value: 0 },
    });
    expect(change).toHaveBeenCalledTimes(0);
    wrapper.find('input').first().simulate('change', {
      target: { value: 5 },
    });
    expect(change).toHaveBeenCalledTimes(1);
    expect(change).toHaveBeenLastCalledWith(5);
  });

  it('wont return values that are too high', () => {
    const change = jest.fn();
    const wrapper = mount(
      <NumberInput value={1} onChange={change} />
    );

    wrapper.find('input').first().simulate('change', {
      target: { value: 999999 },
    });
    expect(change).toHaveBeenCalledTimes(1);
    expect(change).toHaveBeenLastCalledWith(1000);
  });

  it('wont return values above max', () => {
    const change = jest.fn();
    const wrapper = mount(
      <NumberInput value={1} max={100} onChange={change} />
    );

    wrapper.find('input').first().simulate('change', {
      target: { value: 999999 },
    });
    expect(change).toHaveBeenCalledTimes(1);
    expect(change).toHaveBeenLastCalledWith(100);
  });

  it('bubbles enter press', () => {
    const enter = jest.fn();
    const wrapper = mount(
      <NumberInput onEnterPress={enter} />
    );

    expect(enter).toHaveBeenCalledTimes(0);
    wrapper.find('input').first().simulate('keypress', {
      key: 'Enter',
    });
    expect(enter).toHaveBeenCalledTimes(1);
    wrapper.find('input').first().simulate('keypress', {
      key: 'a',
    });
    expect(enter).toHaveBeenCalledTimes(1);
  });
});
