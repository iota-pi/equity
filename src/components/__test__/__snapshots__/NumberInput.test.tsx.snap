// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumberInput/> renders consistently 1`] = `
<div
  class="MuiFormControl-root MuiTextField-root"
>
  <div
    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
  >
    <div
      class="MuiInputAdornment-root MuiInputAdornment-positionStart"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M16 11c1.66 0 2.99-1.34 2.99-3S17.66 5 16 5c-1.66 0-3 1.34-3 3s1.34 3 3 3zm-8 0c1.66 0 2.99-1.34 2.99-3S9.66 5 8 5C6.34 5 5 6.34 5 8s1.34 3 3 3zm0 2c-2.33 0-7 1.17-7 3.5V19h14v-2.5c0-2.33-4.67-3.5-7-3.5zm8 0c-.29 0-.62.02-.97.05 1.16.84 1.97 1.97 1.97 3.45V19h6v-2.5c0-2.33-4.67-3.5-7-3.5z"
        />
      </svg>
    </div>
    <input
      aria-invalid="false"
      class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
      id="player-count"
      min="1"
      step="1"
      type="number"
      value="1"
    />
  </div>
</div>
`;
